proteus/LinearAlgebraTools.py

```diff
@@ -724,8 +983,166 @@ def apply(self,A,x,y):
         -------
         y : vector
         """
+#        _petsc_view(x,'x_mpi_2')
         self.ksp.solve(x,y)
+#        _petsc_view(y,'y_mpi_2')
+#        _petsc_view(self.ksp.getOperators()[0],'Qp_mpi_2')
+#        import pdb ; pdb.set_trace()
+
+class Sp_shell(InvOperatorShell):
+    def __init__(self, A00, A11, A01, A10):
+        self.A00 = A00
+        self.A11 = A11
+        self.A01 = A01
+        self.A10 = A10
+        self._create_Sp()
+        self._options = p4pyPETSc.Options()
+        self._create_constant_nullspace()
+        
+        self.kspSp = p4pyPETSc.KSP().create()
+        self.kspSp.setOperators(self.Sp,self.Sp)
+        self.kspSp.setOptionsPrefix('innerSp_solve_')
+        self.kspSp.setFromOptions()
+        self.Sp.setNullSpace(self.const_null_space)
```

Comment:
Can you explain the route to setting self.const_null_space? Which object has the responsibility of deciding if it has one or not?